client/CPlayerInterface.cpp

```diff
@@ -1685,16 +1686,39 @@ void CPlayerInterface::showShipyardDialogOrProblemPopup(const IShipyard *obj)
 
 void CPlayerInterface::requestReturningToMainMenu(bool won)
 {
+	HighScoreParameter param;
+	param.difficulty = cb->getStartInfo()->difficulty;
+	param.day = cb->getDate();
+	param.townAmount = cb->howManyTowns();
+	param.usedCheat = cb->getPlayerState(*cb->getPlayerID())->enteredWinningCheatCode;
+	param.hasGrail = false;
+	for(const CGHeroInstance * h : cb->getHeroesInfo())
+		if(h->hasArt(ArtifactID::GRAIL))
+			param.hasGrail = true;

```

Comment:
Does not accounts for cases when player builds Grail in town.
Also, not 100% sure, but probably H3 actually check for successful digging action. Meaning, prebuilt grails in towns are not counted, grails acquired in other ways, e.g. picked on map, are not counted either.

Reply:
Checking digging action is also not precise enough, because grail can be obtained by quest for example. There is combined grail mod, which allows to build grail from multiple artifacts.

Reply:
AFAIK grail from quest will not count as "grail" for H3 score calculation

Reply:
Added grail in town